iceoryx_posh/source/popo/ports/server_port_roudi.cpp

```diff
@@ -38,21 +39,94 @@ ServerPortRouDi::MemberType_t* ServerPortRouDi::getMembers() noexcept
     return reinterpret_cast<MemberType_t*>(BasePort::getMembers());
 }
 
+ConsumerTooSlowPolicy ServerPortRouDi::getClientTooSlowPolicy() const noexcept
+{
+    return static_cast<ConsumerTooSlowPolicy>(getMembers()->m_chunkSenderData.m_subscriberTooSlowPolicy);
+}
+
 cxx::optional<capro::CaproMessage> ServerPortRouDi::tryGetCaProMessage() noexcept
 {
-    // nothing to change
-    return cxx::nullopt_t();
+    // get offer state request from user side
+    const auto offeringRequested = getMembers()->m_offeringRequested.load(std::memory_order_relaxed);
+
+    const auto isOffered = getMembers()->m_offered.load(std::memory_order_relaxed);
+
+    if (offeringRequested && !isOffered)
+    {
+        getMembers()->m_offered.store(true, std::memory_order_relaxed);
+
+        capro::CaproMessage caproMessage(capro::CaproMessageType::OFFER, this->getCaProServiceDescription());
+
+        return cxx::make_optional<capro::CaproMessage>(caproMessage);
+    }
+    else if ((!offeringRequested) && isOffered)
+    {
+        getMembers()->m_offered.store(false, std::memory_order_relaxed);
+
+        // remove all the clients (represented by their chunk queues)
+        m_chunkSender.removeAllQueues();
+
+        capro::CaproMessage caproMessage(capro::CaproMessageType::STOP_OFFER, this->getCaProServiceDescription());
+        return cxx::make_optional<capro::CaproMessage>(caproMessage);
+    }
+    else
+    {
+        // nothing to change
+        return cxx::nullopt_t();
+    }
 }
 
 cxx::optional<capro::CaproMessage>
-ServerPortRouDi::dispatchCaProMessageAndGetPossibleResponse(const capro::CaproMessage& /*caProMessage*/) noexcept
+ServerPortRouDi::dispatchCaProMessageAndGetPossibleResponse(const capro::CaproMessage& caProMessage) noexcept
 {
     capro::CaproMessage responseMessage(
         capro::CaproMessageType::NACK, this->getCaProServiceDescription(), capro::CaproMessageSubType::NOSUBTYPE);
 
+    if (!getMembers()->m_offered.load(std::memory_order_relaxed))
+    {
+        return cxx::make_optional<capro::CaproMessage>(responseMessage);
+    }
+
+    switch (caProMessage.m_type)
+    {
+    case capro::CaproMessageType::CONNECT:
+    {
+        cxx::Expects(caProMessage.m_chunkQueueData != nullptr && "Invalid response queue passed to server");
+        const auto ret = m_chunkSender.tryAddQueue(static_cast<ClientChunkQueueData_t*>(caProMessage.m_chunkQueueData),
+                                                   caProMessage.m_historyCapacity);
+        if (!ret.has_error())
+        {
+            responseMessage.m_type = capro::CaproMessageType::ACK;
+            responseMessage.m_chunkQueueData = static_cast<void*>(&getMembers()->m_chunkReceiverData);
+            responseMessage.m_historyCapacity = 0;
+        }
+    }
+    break;
```

Comment:
I would move the break into the scope of the `case`. It feels otherwise a little weird.